internal/provider/aws/provider_test.go

```diff
@@ -0,0 +1,4076 @@
+					})
+
+				status, err := provider.ReconcileFirewallRule(ctx, log, nodeName, instanceID, firewallRule, firewallrules)
+				Expect(status).To(MatchFields(IgnoreExtras, Fields{
+					"State":      Equal(v1alpha1.FirewallRuleStatePending),
+					"Conditions": BeNil(),
+				}))
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("should return a pending state and an error when an AWS API call (CreateSecurityGroup) fails", func() {
+				mockec2Client.EXPECT().
+					DescribeInstances(ctx, gomock.AssignableToTypeOf(&ec2.DescribeInstancesInput{})).
+					DoAndReturn(func(_ context.Context, input *ec2.DescribeInstancesInput, _ ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error) {
+						Expect(input.InstanceIds).To(ConsistOf(instanceID))
+						return &ec2.DescribeInstancesOutput{
+							Reservations: []types.Reservation{
+								{
+									Instances: []types.Instance{
+										{
+											InstanceId: aws.String(instanceID),
+											VpcId:      aws.String(vpcID),
+											NetworkInterfaces: []types.InstanceNetworkInterface{
+												{
+													Association: &types.InstanceNetworkInterfaceAssociation{
+														IpOwnerId: aws.String("aws"),
+														PublicIp:  aws.String(publicIP),
+													},
+													NetworkInterfaceId: aws.String(eniID),
+												},
+											},
+										},
+									},
+								},
+							},
+						}, nil
+					})
+				mockec2Client.EXPECT().
+					DescribeSecurityGroups(ctx, gomock.AssignableToTypeOf(&ec2.DescribeSecurityGroupsInput{})).
+					DoAndReturn(func(_ context.Context, input *ec2.DescribeSecurityGroupsInput, _ ...func(*ec2.Options)) (*ec2.DescribeSecurityGroupsOutput, error) {
+						Expect(input.Filters).To(ConsistOf(filters))
+						return &ec2.DescribeSecurityGroupsOutput{}, nil
+					})
+				mockec2Client.EXPECT().
+					CreateSecurityGroup(ctx, gomock.AssignableToTypeOf(&ec2.CreateSecurityGroupInput{})).
+					DoAndReturn(func(_ context.Context, input *ec2.CreateSecurityGroupInput, _ ...func(*ec2.Options)) (*ec2.CreateSecurityGroupOutput, error) {
+						Expect(aws.ToString(input.VpcId)).To(Equal(vpcID))
+						Expect(input.TagSpecifications[0].ResourceType).To(Equal(types.ResourceTypeSecurityGroup))
+						Expect(input.TagSpecifications[0].Tags).To(ConsistOf(tags))
+						return &ec2.CreateSecurityGroupOutput{}, fmt.Errorf("create security group error")
+					})
+
+				status, err := provider.ReconcileFirewallRule(ctx, log, nodeName, instanceID, firewallRule, firewallrules)
+				Expect(status).To(MatchFields(IgnoreExtras, Fields{
+					"State": Equal(v1alpha1.FirewallRuleStatePending),
+					"Conditions": HaveExactElements(matchCondition(
+						kmetav1.Condition{
+							Type:   v1alpha1.FirewallRuleConditionTypeSecurityGroupCreated,
+							Status: kmetav1.ConditionFalse,
+							Reason: v1alpha1.FirewallRuleConditionReasonProviderError,
+						}, "LastTransitionTime", "ObservedGeneration", "Message")),
+				}))
+				Expect(err).To(HaveOccurred())
```

Comment:
I think it would be interesting to test error types when you assert that an error have occured, e.g. 
```suggestion
			    Expect(err).To(MatchError(&provider.Error{
					Code: "InternalError",
					Msg:  "failed to create security group: create security group error",
				}))
```